src/runtime/parcelset/parcelhandler.cpp

```diff
@@ -462,6 +462,151 @@ namespace hpx { namespace parcelset
         resolver_->route(std::move(p), std::move(wrapped_f));
     }
 
+    void parcelhandler::put_parcels(std::vector<parcel> parcels,
+        std::vector<write_handler_type> handlers)
+    {
+        HPX_ASSERT(resolver_);
+
+        if (parcels.size() != handlers.size())
+        {
+            HPX_THROW_EXCEPTION(bad_parameter,
+                "parcelhandler::put_parcels",
+                "mismatched number of parcels and handlers");
+            return;
+        }
+
+        if (0 == hpx::threads::get_self_ptr())
+        {
+            HPX_ASSERT(!hpx::is_starting());
+
+            naming::gid_type locality = naming::get_locality_from_gid(
+                (*parcels[0].destinations()).get_gid());
+            if (!resolver_->has_resolved_locality(locality))
+            {
+                // reschedule request as an HPX thread to avoid hangs
+                void (parcelhandler::*put_parcels_ptr) (
+                        std::vector<parcel>, std::vector<write_handler_type>
+                    ) = &parcelhandler::put_parcels;
+
+                threads::register_thread_nullary(
+                    util::bind(
+                        util::one_shot(put_parcels_ptr), this,
+                        std::move(parcels), std::move(handlers)),
+                    "parcelhandler::put_parcels", threads::pending, true,
+                    threads::thread_priority_boost);
+                return;
+            }
+        }
+
+        // partition parcels depending on whether their destination can be
+        // resolved locally
+        std::size_t num_parcels = parcels.size();
+
+        std::vector<parcel> resolved_parcels;
+        resolved_parcels.reserve(num_parcels);
+        std::vector<write_handler_type> resolved_handlers;
+        resolved_handlers.reserve(num_parcels);
+
+        typedef std::pair<boost::shared_ptr<parcelport>, locality>
+            destination_pair;
+
+        destination_pair resolved_dest;
+
+        std::vector<parcel> nonresolved_parcels;
+        nonresolved_parcels.reserve(num_parcels);
+        std::vector<write_handler_type> nonresolved_handlers;
+        nonresolved_handlers.reserve(num_parcels);
+
+        for (std::size_t i = 0; i != num_parcels; ++i)
+        {
+            parcel& p = parcels[i];
+
+            // make sure all parcels go to the same locality
+            if (parcels[0].destination_locality() !=
+                p.destination_locality())
+            {
+                HPX_THROW_EXCEPTION(bad_parameter,
+                    "parcelhandler::put_parcels",
+                    "mismatched number of parcels and handlers");
```

Comment:
There's a typo here. It should probably read: "mismatched destinations" or something along the lines.